test_conformance/c11_atomics/common.h

```diff
@@ -948,51 +948,76 @@ int CBasicTest<HostAtomicType, HostDataType>::ExecuteSingleTest(cl_device_id dev
 
   if(deviceThreadCount > 0)
   {
-    cl_ulong usedLocalMemory;
-    cl_ulong totalLocalMemory;
-    cl_uint maxWorkGroupSize;
-
-    // Set up the kernel code
-    programSource = PragmaHeader(deviceID)+ProgramHeader(numDestItems)+FunctionCode()+KernelCode(numDestItems);
-    programLine = programSource.c_str();
-    if (create_single_kernel_helper_with_build_options(
-            context, &program, &kernel, 1, &programLine, "test_atomic_kernel",
-            gOldAPI ? "" : nullptr))
-    {
-      return -1;
-    }
-    if(gDebug)
-    {
-      log_info("Program source:\n");
-      log_info("%s\n", programLine);
-    }
-    // tune up work sizes based on kernel info
-    error = clGetKernelWorkGroupInfo(kernel, deviceID, CL_KERNEL_WORK_GROUP_SIZE, sizeof(groupSize), &groupSize, NULL);
-    test_error(error, "Unable to obtain max work group size for device and kernel combo");
-
-    if(LocalMemory())
-    {
-      error = clGetKernelWorkGroupInfo (kernel, deviceID, CL_KERNEL_LOCAL_MEM_SIZE, sizeof(usedLocalMemory), &usedLocalMemory, NULL);
-      test_error(error, "clGetKernelWorkGroupInfo failed");
-
-      error = clGetDeviceInfo(deviceID, CL_DEVICE_LOCAL_MEM_SIZE, sizeof(totalLocalMemory), &totalLocalMemory, NULL);
-      test_error(error, "clGetDeviceInfo failed");
-
-      // We know that each work-group is going to use typeSize * deviceThreadCount bytes of local memory
-      // so pick the maximum value for deviceThreadCount that uses all the local memory.
-      maxWorkGroupSize = ((totalLocalMemory - usedLocalMemory) / typeSize);
-
-      if(maxWorkGroupSize < groupSize)
-        groupSize = maxWorkGroupSize;
-    }
-
-    CurrentGroupSize((cl_uint)groupSize);
+      // This loop iteratively reduces the workgroup size by 2 and then
+      // re-generates the kernel with the reduced
+      // workgroup size until we find a size which is admissible for the kernel
+      // being run or reduce the wg size
+      // to the trivial case of 1 (which was separately verified to be accurate
+      // for the kernel being run)
+
+      while ((CurrentGroupSize() > 1))
+      {
+          // Re-generate the kernel code with the current group size
+          if (kernel) clReleaseKernel(kernel);
+          if (program) clReleaseProgram(program);
+          programSource = PragmaHeader(deviceID) + ProgramHeader(numDestItems)
+              + FunctionCode() + KernelCode(numDestItems);
+          programLine = programSource.c_str();
+          if (create_single_kernel_helper_with_build_options(
+                  context, &program, &kernel, 1, &programLine,
+                  "test_atomic_kernel", gOldAPI ? "" : "-cl-std=CL2.0"))
```

Comment:
Was it intentional to go back to specifying `-cl-std=CL2.0`?

Reply:
oops no, that was a typo. Thanks for catching it, I'll update.

Reply:
@Jeremy-Kemp can you review again? Thanks.